### PR TITLE
datanode's dn_rpc_port supports ssl

### PR DESCRIPTION
**Pull request by CloudWise-Lukemiao**
datanode's dn_rpc_port supports ssl

### CODE REVIEW

File: iotdb-client/cli/src/main/java/org/apache/iotdb/cli/Cli.java [hunk 1]
```diff
@@ -137,17 +141,25 @@ private static void serve() {
         password = lineReader.readLine("please input your password:", '\0');
       }
       receiveCommands(lineReader);
+
```
Comment (OneSizeFitsQuorum) [inline iotdb-client/cli/src/main/java/org/apache/iotdb/cli/Cli.java:144]: remove empty line

File: iotdb-client/cli/src/main/java/org/apache/iotdb/cli/Cli.java [hunk 2]
```diff
@@ -137,17 +141,25 @@ private static void serve() {
...
     } catch (Exception e) {
       println(IOTDB_ERROR_PREFIX + ": Exit cli with error: " + e.getMessage());
       System.exit(CODE_ERROR);
     }
   }
 
   private static void executeSql() throws TException {
+    Properties info = new Properties();
```
Comment (OneSizeFitsQuorum) [inline iotdb-client/cli/src/main/java/org/apache/iotdb/cli/Cli.java:152]: make it a class field?

File: iotdb-client/cli/src/main/java/org/apache/iotdb/cli/Cli.java [hunk 3]
```diff
@@ -161,10 +173,17 @@ private static void executeSql() throws TException {
   }
 
   private static void receiveCommands(LineReader reader) throws TException {
+    Properties info = new Properties();
```
Comment (OneSizeFitsQuorum) [inline iotdb-client/cli/src/main/java/org/apache/iotdb/cli/Cli.java:176]: make it a class field?

File: iotdb-client/jdbc/src/main/java/org/apache/iotdb/jdbc/IoTDBConnection.java [hunk 4]
```diff
@@ -469,11 +469,22 @@ public void setClient(IClientRPCService.Iface client) {
...
+              params.getHost(),
+              params.getPort(),
+              getNetworkTimeout(),
+              params.getTrustStore(),
+              params.getTrustStorePwd());
+    } else {
+      transport =
+          RpcTransportFactory.INSTANCE.getTransport(
+              params.getHost(), params.getPort(), getNetworkTimeout());
+      if (!transport.isOpen()) {
```
Comment (OneSizeFitsQuorum) [inline iotdb-client/jdbc/src/main/java/org/apache/iotdb/jdbc/IoTDBConnection.java:485]: move it to line 488?

File: iotdb-client/session/src/main/java/org/apache/iotdb/session/Session.java [hunk 5]
```diff
@@ -316,6 +319,230 @@ public Session(
...
+        SessionConfig.DEFAULT_INITIAL_BUFFER_CAPACITY,
+        SessionConfig.DEFAULT_MAX_FRAME_SIZE,
+        SessionConfig.DEFAULT_REDIRECTION_MODE,
+        SessionConfig.DEFAULT_VERSION,
+        useSSL,
+        trustStore,
+        trustStorePwd);
+  }
+
+  public Session(
```
Comment (OneSizeFitsQuorum) [inline iotdb-client/session/src/main/java/org/apache/iotdb/session/Session.java:340]: no need to add so many new constructor.

File: iotdb-client/session/src/main/java/org/apache/iotdb/session/Session.java [hunk 6]
```diff
@@ -3536,19 +3895,34 @@ public Session build() {
                 version);
         newSession.setEnableQueryRedirection(true);
         return newSession;
+      } else if (nodeUrls == null && useSSL) {
```
Comment (OneSizeFitsQuorum) [inline iotdb-client/session/src/main/java/org/apache/iotdb/session/Session.java:3898]: please still use the same two branches as before, and don't have a separate if else branch for ssl

File: iotdb-client/session/src/main/java/org/apache/iotdb/session/SessionConnection.java [hunk 7]
```diff
@@ -118,17 +118,29 @@ public SessionConnection(Session session, ZoneId zoneId) throws IoTDBConnectionE
...
+                endPoint.getPort(),
+                session.connectionTimeoutInMs,
+                trustStore,
+                trustStorePwd);
+      } else {
+        transport =
+            RpcTransportFactory.INSTANCE.getTransport(
+                // as there is a try-catch already, we do not need to use TSocket.wrap
+                endPoint.getIp(), endPoint.getPort(), session.connectionTimeoutInMs);
+        if (!transport.isOpen()) {
```
Comment (OneSizeFitsQuorum) [inline iotdb-client/session/src/main/java/org/apache/iotdb/session/SessionConnection.java:139]: move to line 142

File: iotdb-client/session/src/main/java/org/apache/iotdb/session/SessionConnection.java [hunk 8]
```diff
@@ -118,17 +118,29 @@ public SessionConnection(Session session, ZoneId zoneId) throws IoTDBConnectionE
...
+          transport.open();
+        }
       }
+
```
Comment (OneSizeFitsQuorum) [inline iotdb-client/session/src/main/java/org/apache/iotdb/session/SessionConnection.java:143]: remove empty line

File: iotdb-client/session/src/main/java/org/apache/iotdb/session/pool/SessionPool.java [hunk 9]
```diff
@@ -278,6 +429,35 @@ public SessionPool(
         SessionConfig.DEFAULT_MAX_FRAME_SIZE);
   }
 
+  public SessionPool(
```
Comment (OneSizeFitsQuorum) [inline iotdb-client/session/src/main/java/org/apache/iotdb/session/pool/SessionPool.java:432]: same as above

File: iotdb-client/session/src/main/java/org/apache/iotdb/session/pool/SessionPool.java [hunk 10]
```diff
@@ -404,6 +725,40 @@ private Session constructNewSession() {
               .thriftMaxFrameSize(thriftMaxFrameSize)
               .enableRedirection(enableRedirection)
               .version(version)
+              .useSSL(useSSL)
+              .trustStore(trustStore)
+              .trustStorePwd(trustStorePwd)
+              .build();
+    } else if (nodeUrls == null && !useSSL) {
```
Comment (OneSizeFitsQuorum) [inline iotdb-client/session/src/main/java/org/apache/iotdb/session/pool/SessionPool.java:732]: same as above

File: iotdb-client/session/src/main/java/org/apache/iotdb/session/pool/SessionPool.java [hunk 11]
```diff
@@ -3533,6 +3907,43 @@ public SessionPool build() {
             version,
             thriftDefaultBufferSize,
             thriftMaxFrameSize);
+      } else if (nodeUrls == null && useSSL) {
```
Comment (OneSizeFitsQuorum) [inline iotdb-client/session/src/main/java/org/apache/iotdb/session/pool/SessionPool.java:3910]: same as above

File: iotdb-core/datanode/src/assembly/resources/conf/iotdb-datanode.properties [hunk 12]
```diff
@@ -60,6 +60,19 @@ dn_data_region_consensus_port=10760
...
+# enable_thrift_ssl=false
+
+# SSL key store path
+# key_store_path=
+
+# SSL key store password
+# key_store_pwd=
+
+# SSL client timeout (in millisecond)
+# client_timeout=300000
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/datanode/src/assembly/resources/conf/iotdb-datanode.properties:74]: What's the point of separating out a single parameter? Is there a problem if we use dn_connection_timeout?

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java [hunk 13]
```diff
@@ -124,6 +124,18 @@ public class IoTDBConfig {
   /** Port which the JDBC server listens to. */
   private int rpcPort = 6667;
 
+  /** Enable the thrift rpcPort Service ssl. */
+  private boolean enableSSL = false;
+
+  /** ssl key Store Path. */
+  private String keyStorePath = "";
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java:131]: Initialize the defaults elsewhere to "" as well.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java [hunk 14]
```diff
@@ -124,6 +124,18 @@ public class IoTDBConfig {
...
+
+  /** ssl key Store password. */
+  private String keyStorePwd = "";
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java:134]: same

**Review by OneSizeFitsQuorum — COMMENTED** [review]
Awesome work! 
I have left some issues, PTAL

File: iotdb-client/cli/src/main/java/org/apache/iotdb/cli/Cli.java [hunk 15]
```diff
@@ -161,10 +174,16 @@ private static void executeSql() throws TException {
   }
 
   private static void receiveCommands(LineReader reader) throws TException {
+    if (useSsl != null && Boolean.parseBoolean(useSsl)) {
```
Comment (OneSizeFitsQuorum) [inline iotdb-client/cli/src/main/java/org/apache/iotdb/cli/Cli.java:177]: same

File: iotdb-client/session/src/main/java/org/apache/iotdb/session/Session.java [hunk 16]
```diff
@@ -3521,34 +3562,11 @@ public Session build() {
...
-                nodeUrls,
-                username,
-                pw,
-                fetchSize,
-                zoneId,
-                thriftDefaultBufferSize,
-                thriftMaxFrameSize,
-                enableRedirection,
-                version);
         newSession.setEnableQueryRedirection(true);
```
Comment (OneSizeFitsQuorum) [inline iotdb-client/session/src/main/java/org/apache/iotdb/session/Session.java:3567]: move to constructor?

File: iotdb-client/session/src/main/java/org/apache/iotdb/session/SessionConnection.java (hunk 7 above)
Comment (OneSizeFitsQuorum) [reply iotdb-client/session/src/main/java/org/apache/iotdb/session/SessionConnection.java:139]: move to line 142

File: iotdb-client/session/src/main/java/org/apache/iotdb/session/SessionConnection.java (hunk 8 above)
Comment (OneSizeFitsQuorum) [reply iotdb-client/session/src/main/java/org/apache/iotdb/session/SessionConnection.java:143]: remove empty line

File: iotdb-client/session/src/main/java/org/apache/iotdb/session/pool/SessionPool.java [hunk 17]
```diff
@@ -404,6 +484,40 @@ private Session constructNewSession() {
...
+              .password(password)
+              .fetchSize(fetchSize)
+              .zoneId(zoneId)
+              .thriftDefaultBufferSize(thriftDefaultBufferSize)
+              .thriftMaxFrameSize(thriftMaxFrameSize)
+              .enableRedirection(enableRedirection)
+              .version(version)
+              .build();
+
+    } else if (nodeUrls != null && useSSL) {
```
Comment (OneSizeFitsQuorum) [inline iotdb-client/session/src/main/java/org/apache/iotdb/session/pool/SessionPool.java:506]: I feel like we can brute force the new parameters to be set in, without the useSSL judgment?

File: iotdb-core/datanode/src/assembly/resources/conf/iotdb-datanode.properties [hunk 18]
```diff
@@ -254,4 +267,44 @@ dn_target_config_node_list=127.0.0.1:10710
 # The type of internal reporter in metric module, used for checking flushed point number
 # Options: [MEMORY, IOTDB]
 # Datatype: String
-# dn_metric_internal_reporter_type=MEMORY
\ No newline at end of file
+# dn_metric_internal_reporter_type=MEMORY
+
+####################
+### REST Service Configuration
```
Comment (OneSizeFitsQuorum) [inline iotdb-core/datanode/src/assembly/resources/conf/iotdb-datanode.properties:273]: need to update doc

**Review by OneSizeFitsQuorum — APPROVED** [review]
LGTM~
BTW, It is better to add some performance test results in the pr description.

**Comment by CloudWise-Lukemiao** [comment]
> LGTM~ BTW, It is better to add some performance test results in the pr description.

Using the insertTablet method of the session, using 100 devices * 100 test points, each batch of 100 rows is cycled 100 times, tested 3 times, with three insertions using different device names. Test environment: 1c1d, Mac 16G, Intel Core i5. Test results: Based on the average of the three insertions of SSL, the average value is 13117ms, and the average value of non SSL is 12666ms. Overall, the test results show that the insertion speed of non SSL is slightly faster than that of SSL。

Using the executeQueryStatement method of the session, loop 100 times to query the test points of different devices, test 3 times, modify the SSL after the query, restart Iotdb, and use different device names for the three queries. Test environment: 1c1d, Mac 16G, Intel Core i5. Test conclusion: Based on the average of 4601ms for SSL three times and 3768ms for non SSL, the comprehensive test results show that the insertion speed of non SSL is slightly faster for SSL

**Pull request merged**